### ISSUE TITLE
Bleadperl v5.19.1-480-g9baac1a breaks COWENS/autobox-dump-20090426.1746.tar.gz, APLA/Project-Easy-0.30.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13113]
Migrated from [rt.perl.org#118933](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=118933) (status was 'resolved')

Searchable as RT118933$

**Comment by p5pRT** [comment]
### From @andk

git bisect


---
commit 9baac1a3613bd641a847683d7877b3cfab3244bc
Author&#8203;: Slaven Rezic \<srezic@&#8203;iconmobile\.com>
Date&#8203;:   Wed Jul 10 14&#8203;:18&#8203;:18 2013 \+1000

&nbsp;    Data&#8203;::Dumper&#8203;: useqq implementation for xs

diagnostics


---
http&#8203;://www\.cpantesters\.org/cpan/report/9054ada6\-eefd\-11e2\-8c2c\-94abf1ff63fb

No new enough reports yet available on cpantesters for autobox\-dump\.

perl \-V


---
Summary of my perl5 \(revision 5 version 19 subversion 2\) configuration&#8203;:      
&nbsp;  Commit id&#8203;: 9baac1a3613bd641a847683d7877b3cfab3244bc
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.9\-1\-amd64\,
&nbsp;  archname=x86\_64\-linux\-thread\-multi\-ld
&nbsp;    uname='linux k83 3\.9\-1\-amd64 \#1 smp debian 3\.9\.8\-1 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dprefix=/home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.1\-480\-g9baac1a/a2da
&nbsp;  \-Dmyhostname=k83 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des
&nbsp;  \-Ui\_db \-Duseithreads \-Duselongdouble \-DDEBUGGING=\-g'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=define
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing
&nbsp;  \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
&nbsp;  \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-pipe
&nbsp;  \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.8\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\,
&nbsp;  Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=16\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib
&nbsp;  /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;  \-lgdbm\_compat
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.17'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib
&nbsp;  \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: HAS\_TIMES MULTIPLICITY PERLIO\_LAYERS
&nbsp;                        PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_HASH\_FUNC\_ONE\_AT\_A\_TIME\_HARD
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_NEW\_COPY\_ON\_WRITE PERL\_PRESERVE\_IVUV
&nbsp;                        PERL\_USE\_DEVEL USE\_64\_BIT\_ALL USE\_64\_BIT\_INT
&nbsp;                        USE\_ITHREADS USE\_LARGE\_FILES USE\_LOCALE
&nbsp;                        USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_LONG\_DOUBLE USE\_PERLIO
&nbsp;                        USE\_PERL\_ATOF USE\_REENTRANT\_API
&nbsp;  Built under linux
&nbsp;  Compiled at Jul 19 2013 04&#8203;:32&#8203;:06
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.1\-480\-g9baac1a/a2da/lib/site\_perl/5\.19\.2/x86\_64\-linux\-thread\-multi\-ld
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.1\-480\-g9baac1a/a2da/lib/site\_perl/5\.19\.2
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.1\-480\-g9baac1a/a2da/lib/5\.19\.2/x86\_64\-linux\-thread\-multi\-ld
&nbsp;    /home/sand/src/perl/repoperls/installed\-perls/perl/v5\.19\.1\-480\-g9baac1a/a2da/lib/5\.19\.2
&nbsp;    \.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @tonycoz

On Fri Jul 19 02&#8203;:50&#8203;:51 2013\, andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de wrote&#8203;:
> git bisect
> \-\-\-\-\-\-\-\-\-\-
> commit 9baac1a3613bd641a847683d7877b3cfab3244bc
> Author&#8203;: Slaven Rezic \<srezic@&#8203;iconmobile\.com>
> Date&#8203;:   Wed Jul 10 14&#8203;:18&#8203;:18 2013 \+1000
> 
>     Data&#8203;::Dumper&#8203;: useqq implementation for xs
> 
> diagnostics
> \-\-\-\-\-\-\-\-\-\-\-
> http&#8203;://www\.cpantesters\.org/cpan/report/9054ada6\-eefd\-11e2\-8c2c\-
> 94abf1ff63fb

The failure in Project&#8203;::Easy is from it fiddling directly with the
internals of Data&#8203;::Dumper\.

Project&#8203;::Easy&#8203;::Config&#8203;::Format&#8203;::perl contains the code&#8203;:

\{
&nbsp;    no warnings 'redefine';
&nbsp;    sub Data&#8203;::Dumper&#8203;::qquote \{
&nbsp;        my $s = shift;
&nbsp;        return "'$s'";
&nbsp;    \}
\}

and then forces Useqq later to try to force the perl Dump\(\) to be used\.

Since the XS dumper doesn't call sub qquote\(\) the replacement doesn't occur\.

autobox\-dump looks like a different problem\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Jul 21 18&#8203;:28&#8203;:32 2013\, tonyc wrote&#8203;:
> autobox\-dump looks like a different problem\.

Which it is\.

The pure\-perl dumper \(with or without useqq\) outputs SVs which can be
represented as small\(\-ish\) decimal numbers \*as\* numbers\, but the new
useqq handling didn't\.

The XS non\-useqq code however doesn't do this\.  rather than probably
breaking many more CPAN tests by making them consistent I'm adding code
that does the same check as the pure perl code does\, but enabling it
only for useqq\.

The attached patch\, which I'll apply post 5\.19\.2 \(and probably with a
version bump\) allows autobox&#8203;::dump to pass\.

I don't plan on any changes to support Project&#8203;::Easy\, that's purely a
bug in that module\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Jul 21 22&#8203;:31&#8203;:54 2013\, tonyc wrote&#8203;:
> The attached patch\, which I'll apply post 5\.19\.2 \(and probably with a
> version bump\) allows autobox&#8203;::dump to pass\.

Now with more attachment\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-74798-improved-useqq-compatibility-with-the-pur.patch</summary>

```diff
From f8e039f96e269609b4b496ca3fc8f15b1ea8c393 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Mon, 22 Jul 2013 14:59:57 +1000
Subject: [PATCH] [perl #74798] improved useqq compatibility with the pure
 perl version

Currently for non-useqq, the pure perl and XS output for numbers like
these is different, but XS useqq is new, so try to remain vaguely
compatible.
---
 dist/Data-Dumper/Dumper.xs  |   45 +++++++++++++++++++++++++++++++++++++++++++
 dist/Data-Dumper/t/dumper.t |   21 ++++++++++++++++++--
 2 files changed, 64 insertions(+), 2 deletions(-)

diff --git a/dist/Data-Dumper/Dumper.xs b/dist/Data-Dumper/Dumper.xs
index 0194a2c..3bcb2f4 100644
--- a/dist/Data-Dumper/Dumper.xs
+++ b/dist/Data-Dumper/Dumper.xs
@@ -119,6 +119,42 @@ TOP:
     return 0;
 }
 
+/* Check that the SV can be represented as a simple decimal integer.
+ *
+ * The perl code does this by matching against /^(?:0|-?[1-9]\d{0,8})\z/
+*/
+static I32
+safe_decimal_number(SV *val) {
+    STRLEN len;
+    const char *p = SvPV(val, len);
+
+    if (len == 1 && *p == '0')
+        return TRUE;
+
+    if (len && *p == '-') {
+        ++p;
+        --len;
+    }
+
+    if (len == 0 || *p < '1' || *p > '9')
+        return FALSE;
+
+    ++p;
+    --len;
+
+    if (len > 8)
+        return FALSE;
+
+    while (len > 0) {
+         /* the perl code checks /\d/ but we don't want unicode digits here */
+         if (*p < '0' || *p > '9')
+             return FALSE;
+         ++p;
+         --len;
+    }
+    return TRUE;
+}
+
 /* count the number of "'"s and "\"s in string */
 static I32
 num_q(const char *s, STRLEN slen)
@@ -1115,6 +1151,15 @@ DD_dump(pTHX_ SV *val, const char *name, STRLEN namelen, SV *retval, HV *seenhv,
 	    sv_catpvn(retval, (const char *)mg->mg_ptr, mg->mg_len);
 	}
 #endif
+
+	/* the pure perl and XS non-qq outputs have historically been
+	 * different in this case, but for useqq, let's try to match
+	 * the pure perl code.
+	 * see [perl #74798]
+	 */
+	else if (useqq && safe_decimal_number(val)) {
+	    sv_catsv(retval, val);
+	}
 	else {
         integer_came_from_string:
 	    c = SvPV(val, i);
diff --git a/dist/Data-Dumper/t/dumper.t b/dist/Data-Dumper/t/dumper.t
index 0a3c28c..dbc6d5e 100644
--- a/dist/Data-Dumper/t/dumper.t
+++ b/dist/Data-Dumper/t/dumper.t
@@ -83,11 +83,11 @@ sub SKIP_TEST {
 $Data::Dumper::Useperl = 1;
 if (defined &Data::Dumper::Dumpxs) {
   print "### XS extension loaded, will run XS tests\n";
-  $TMAX = 420; $XS = 1;
+  $TMAX = 426; $XS = 1;
 }
 else {
   print "### XS extensions not loaded, will NOT run XS tests\n";
-  $TMAX = 210; $XS = 0;
+  $TMAX = 213; $XS = 0;
 }
 
 print "1..$TMAX\n";
@@ -1555,4 +1555,21 @@ EOW
     "\\ octal followed by unicode digit";
   TEST q(Data::Dumper->Dumpxs(["\\x00\\x{0660}"])), '\\ octal followed by unicode digit (xs)'
     if $XS;
+
+  # [perl #118933 - handling of digits
+$WANT = <<'EOW';
+#$VAR1 = 0;
+#$VAR2 = 1;
+#$VAR3 = 90;
+#$VAR4 = -10;
+#$VAR5 = "010";
+#$VAR6 = 112345678;
+#$VAR7 = "1234567890";
+EOW
+  TEST q(Data::Dumper->Dump([0, 1, 90, -10, "010", "112345678", "1234567890" ])),
+    "numbers and number-like scalars";
+
+  TEST q(Data::Dumper->Dumpxs([0, 1, 90, -10, "010", "112345678", "1234567890" ])),
+    "numbers and number-like scalars"
+    if $XS;
 }
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Jul 21\, 2013 at 10&#8203;:34&#8203;:22PM \-0700\, Tony Cook via RT wrote&#8203;:


> \+/\* Check that the SV can be represented as a simple decimal integer\.
> \+ \*
> \+ \* The perl code does this by matching against /^\(?&#8203;:0|\-?\[1\-9\]\\d\{0\,8\}\)\\z/
> \+\*/
> \+static I32
> \+safe\_decimal\_number\(SV \*val\) \{

> \+        return TRUE;

> \+        return FALSE;

> \+        return FALSE;

> \+             return FALSE;

> \+    return TRUE;
> \+\}


Wouldn't that be better as static BOOL?
Fewer I32s\, not more\, surely? :\-\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon\, Jul 22\, 2013 at 08&#8203;:01&#8203;:34AM \+0100\, Nicholas Clark wrote&#8203;:
> On Sun\, Jul 21\, 2013 at 10&#8203;:34&#8203;:22PM \-0700\, Tony Cook via RT wrote&#8203;:
> 
> 
> > \+/\* Check that the SV can be represented as a simple decimal integer\.
> > \+ \*
> > \+ \* The perl code does this by matching against /^\(?&#8203;:0|\-?\[1\-9\]\\d\{0\,8\}\)\\z/
> > \+\*/
> > \+static I32
> > \+safe\_decimal\_number\(SV \*val\) \{
> 
> > \+        return TRUE;
> 
> > \+        return FALSE;
> 
> > \+        return FALSE;
> 
> > \+             return FALSE;
> 
> > \+    return TRUE;
> > \+\}
> 
> 
> Wouldn't that be better as static BOOL?
> Fewer I32s\, not more\, surely? :\-\)

I'll fix that\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Jul 22 02&#8203;:57&#8203;:17 2013\, tonyc wrote&#8203;:
> On Mon\, Jul 22\, 2013 at 08&#8203;:01&#8203;:34AM \+0100\, Nicholas Clark wrote&#8203;:
> > Wouldn't that be better as static BOOL?
> > Fewer I32s\, not more\, surely? :\-\)
> 
> I'll fix that\.

Applied with fix as 059639d5cdd8e4ce8732e497e1c8a0d9deafc7b3\.

I've opened a ticket against Project&#8203;::Easy&#8203;:

https://rt.cpan.org/Ticket/Display.html?id=87171

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Jul 22 16&#8203;:54&#8203;:17 2013\, tonyc wrote&#8203;:
> On Mon Jul 22 02&#8203;:57&#8203;:17 2013\, tonyc wrote&#8203;:
> > On Mon\, Jul 22\, 2013 at 08&#8203;:01&#8203;:34AM \+0100\, Nicholas Clark wrote&#8203;:
> > > Wouldn't that be better as static BOOL?
> > > Fewer I32s\, not more\, surely? :\-\)
> > 
> > I'll fix that\.
> 
> Applied with fix as 059639d5cdd8e4ce8732e497e1c8a0d9deafc7b3\.
> 
> I've opened a ticket against Project&#8203;::Easy&#8203;:
> 
> https://rt.cpan.org/Ticket/Display.html?id=87171

Since I've fixed the problem causing autobox\-dump to fail\, and reported
the issue in Project&#8203;::Easy\, I'm closing this ticket\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
